FEA Utility function to add arbitrary files to the repo

BenjaminBossan opened the pull request:
Solves #111 

[This](https://huggingface.co/skops-ci/hf_hub_example-c57afd8c-0f9f-4dcc-9fd0-29325a03cf83/tree/main) is the outcome of using `plot_hf_hub.py` with one added line:

```python
hub_utils.add_files([__file__], dst=local_repo)
```

As can be seen, the script itself is now included in the upload.

BenjaminBossan commented:
@skops-dev/maintainers 

Could you help me figure out why the new tests are failing? The offending code seems to stem from this fixutre:

https://github.com/BenjaminBossan/skops/blob/adding-files/skops/hub_utils/tests/test_hf_hub.py#L497-L510

which results in the error `OSError: Model file '/tmp/.../model.pickle' does not exist.`

This seems odd to me because the tests pass locally and because I'm doing basically exactly the same as, e.g., [this test](https://github.com/skops-dev/skops/blob/ca853d4d49c64484e5dcd7c49e1bc03e68b1a0ba/skops/hub_utils/tests/test_hf_hub.py#L242-L254), which succeeds.

BenjaminBossan commented:
@skops-dev/maintainers The PR is ready for review now.

The error mentioned above was caused by `test_inference` doing some cleanup that would disrupt tests running later. We use some session-scoped fixtures, resulting in this dependency. My solution was to have `test_inference` run in a different directory in order for it not to interfere with other tests. Alternatively, we could reduce the scope of the used fixtures, but that would slow the tests down.

Regarding the failed codecov check: I have covered all new lines with tests, so this is a false positive. It seems that codecov thinks that my changes affect the coverage of `get_model_output` (see very last line):

https://app.codecov.io/gh/skops-dev/skops/compare/123/tree/skops/hub_utils/_hf_hub.py#D1L634

This is probably caused by this CI run not randomly failing when calling inference API, so it's a fluke.

adrinjalali commented on docs/changes.rst:
```diff
@@ -9,6 +9,11 @@ skops Changelog
 
+v0.3
+----
+- Utility to function to add arbitrary files to be uploaded to the hub by using
```
```suggestion
- Utility function to add arbitrary files to be uploaded to the hub by using
```

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +328,65 @@ def init(
         raise
 
 
+def add_files(files: Sequence[str | Path], dst: str | Path) -> None:
```
we could also have the signature as:

```suggestion
def add_files(*files, dst):
```

makes it easier for people to just add many while adding one would be easier as:

```python
add_files(__file__, dst)
```

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +328,65 @@ def init(
+
+    # validate input types
+    if isinstance(files, (str, Path)):
+        msg = (
+            "First argument must be a sequence of str or Path, got "
+            f"{type(files)} instead."
+        )
+        raise TypeError(msg)
```
even if we stick with the current signature, I think we should accept a simple string or Path here.

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +328,65 @@ def init(
+
+    dst_files = [dst / Path(file).name for file in files]
+    for src_file, dst_file in zip(src_files, dst_files):
+        if dst_file.exists():
+            msg = f"File '{src_file.name}' already found at '{dst}', skipping."
+            warnings.warn(msg)
+            continue
```
we should have an argument defining the behavior on existing files.

adrinjalali reviewed (commented):
first look, haven't checked the tests yet.

BenjaminBossan commented:
@adrinjalali Yes, I was thinking about making files `*args` too, I changed it as suggested. Furthermore, I added an `exist_ok` argument (default `True`). If the file exists and `exist_ok=True`, I'll now overwrite it (instead of skipping like previously) and if `exist_ok=False` raise an error.

I was also thinking about adding directories but if we want that, let's do it in another PR.

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +327,62 @@ def init(
+    Parameters
+    ----------
+    *files : str or Path
+        The files to be added. If a file by that name already exists at target
+        location, it will be skipped.
```
```suggestion
        The files to be added.
```

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +327,62 @@ def init(
+        :func:`.hub_utils.init`.
+
+    exist_ok : bool (default=True)
+        Whether it's okay or not to add a file that already exists. If it's
+        okay, override the file, otherwise raise a ``FileExistsError``.
```
```suggestion
        Whether it's okay or not to add a file that already exists. If
        ``True``, override the files, otherwise raise a ``FileExistsError``.
```

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +327,62 @@ def init(
+        Path to the initialized repo, same as used during
+        :func:`.hub_utils.init`.
+
+    exist_ok : bool (default=True)
```
Probably set the default to ``False``, to avoid accidental overwrite?

BenjaminBossan replied:
I think I slightly prefer the default to be `True`. That way, calling it multiple times is fine (idempotency).

BenjaminBossan replied:
Not really sure if there is a Python convention for `exist_ok`: https://docs.python.org/3/library/pathlib.html#pathlib.Path.touch

Anyway, I changed it, as mentioned it was only a slight preference.

adrinjalali commented on skops/hub_utils/_hf_hub.py:
```diff
@@ -327,6 +327,62 @@ def init(
+        Whether it's okay or not to add a file that already exists. If it's
+        okay, override the file, otherwise raise a ``FileExistsError``.
+
+    Raises
```
nit: `copy2` might also raise due to permission issues, but I'm okay not listing them here.

BenjaminBossan replied:
When the function is used elsewhere, we also don't mention that error. Not sure if we should document any possible error the function could raise or only those we explicitly raise in the function body?

adrinjalali reviewed (commented):
Otherwise LGTM.
